collect_app/src/androidTest/java/org/odk/collect/android/benchmark/EntitiesBenchmarkTest.kt

```diff
@@ -0,0 +1,137 @@
+import org.odk.collect.android.test.BuildConfig.ENTITIES_FILTER_TEST_PROJECT_URL
+import org.odk.collect.strings.R
+
+/**
+ * Benchmarks the performance of entity follow up forms. [PROJECT_URL] should be set to a project
+ * that contains the "100k Entities Filter" form.
+ *
+ * Devices that currently pass:
+ * - Pixel 4a
+ * - Fairphone 3
+ *
+ */
+
+@RunWith(AndroidJUnit4::class)
+class EntitiesBenchmarkTest {
+
+    private val rule = CollectTestRule(useDemoProject = false)
+
+    @get:Rule
+    var chain: RuleChain = chain(TestDependencies(true)).around(rule)
+
+    @Test
+    fun run() {
+        assertThat(
+            "Need to set ENTITIES_FILTER_TEST_PROJECT_URL before running!",
+            ENTITIES_FILTER_TEST_PROJECT_URL,
+            not(blankOrNullString())
+        )
+        clearAndroidCache()
+
+        val stopwatch = Stopwatch()
+
+        rule.startAtFirstLaunch()
+            .clickManuallyEnterProjectDetails()
+            .inputUrl(ENTITIES_FILTER_TEST_PROJECT_URL)
+            .addProject()
+
+            // Populate http cache and clear out form/entities
+            .clickGetBlankForm()
+            .clickGetSelected()
+            .clickOK(MainMenuPage())
+            .openProjectSettingsDialog()
+            .clickSettings()
+            .clickProjectManagement()
+            .clickOnResetProject()
+            .clickOnString(R.string.reset_blank_forms)
+            .clickOnString(R.string.reset_saved_forms)
+            .clickOnString(R.string.reset_settings_button_reset)
+            .clickOKOnDialog(MainMenuPage())
+
+            .clickGetBlankForm()
+            .benchmark("Downloading form with http cache", stopwatch) {
+                it.clickGetSelected()
+            }
+
+            .clickOK(MainMenuPage())
+            .clickGetBlankForm()
+            .benchmark("Updating form with http cache", stopwatch) {
+                it.clickGetSelected()
+            }
+
+            .clickOK(MainMenuPage())
+            .clickFillBlankForm()
+            .benchmark("Loading form first time", stopwatch) {
+                it.clickOnForm("100k Entities Filter")
+            }
+
+            .pressBackAndDiscardForm()
+            .clickFillBlankForm()
+            .benchmark("Loading form second time", stopwatch) {
+                it.clickOnForm("100k Entities Filter")
+            }
+
+            .answerQuestion("Which value do you want to filter by?", "1024")
+            .benchmark("Filtering select", stopwatch) {
+                it.swipeToNextQuestion("Filtered select")
+            }
+
+        assertThat(stopwatch.getTime("Downloading form with http cache"), lessThan(75))
+        assertThat(stopwatch.getTime("Updating form with http cache"), lessThan(90))
+        assertThat(stopwatch.getTime("Loading form first time"), lessThan(5))
+        assertThat(stopwatch.getTime("Loading form second time"), lessThan(5))
+        assertThat(stopwatch.getTime("Filtering select"), lessThan(5))
```

Comment:
I think you can add const for those keys as they are used here and above.

Reply:
I've changed the way we benchmark/assert so that they only need to be referenced once instead.